### ISSUE TITLE
`float` member issue with `clang++-15` and `g++-12`

### ISSUE THREAD

**Issue by t-mat** [stephenberry/glaze#1007]
Hi, perhaps this is a compiler side issue, but I can't compile the basic structure, which contains `float` with `clang++-15` or `g++-12`.

Structure: `struct MyStruct { float f; };`
Actual code as a gist: https://gist.github.com/t-mat/7869e34b1eab346aa31176f3b27a39c8

Reproducing procedure:

```
cd
git clone https://github.com/stephenberry/glaze glaze-issue-float
cd glaze-issue-float
git log -1
# commit 1586097742ef3cb04ab5dbda15173cb75cc65075 (HEAD -> main, tag: v2.6.4, origin/main, origin/HEAD)
# Author: Stephen Berry <stephenberry.developer@gmail.com>
# Date:   Fri May 17 13:28:42 2024 -0500
# 
#     version 2.6.4 bump

curl -LOJ https://gist.githubusercontent.com/t-mat/7869e34b1eab346aa31176f3b27a39c8/raw/my-glaze-test-float.cpp

clang++-15 -std=c++20 -I ./include ./my-glaze-test-float.cpp
# Error
clang++-15 --version
# Ubuntu clang version 15.0.7

g++-12 -std=c++20 -I ./include ./my-glaze-test-float.cpp
# Error
g++-12 --version
# g++-12 (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
```

<details><summary>Actual code</summary>

```C++
// my-glaze-test-float.cpp

#if defined(__GNUC__)
#  pragma GCC diagnostic push
#  pragma GCC diagnostic ignored "-Wsign-conversion"
#  pragma GCC diagnostic ignored "-Wshadow"
#  pragma GCC diagnostic ignored "-Wconversion"
#endif
#include <glaze/glaze.hpp>
#if defined(__GNUC__)
#  pragma GCC diagnostic pop
#endif
#include <iostream>

struct MyStruct {
    float f; // NG
};

std::vector<uint8_t> serialize(const MyStruct &myStruct)
{
    std::vector<uint8_t> buf;
    glz::write_json(myStruct, buf);
    buf.push_back('\0');
    return buf;
}

std::optional<MyStruct> deserialize(std::vector<uint8_t> &&stream)
{
    MyStruct myStruct;
    if (auto const error = glz::read_json(myStruct, stream)) {
        return std::nullopt;
    }
    return myStruct;
}

int main()
{
    MyStruct myStruct;

    auto const ser = serialize(myStruct);

    std::string str{ reinterpret_cast<const char *>(ser.data()), ser.size() };

    std::cout << str << "\n";
}
```

</details>

<details><summary>Actual error log (`clang++-15`)</summary>

```
~/glaze-issue-float$ clang++-15 -std=c++20 -I ./include ./my-glaze-test-float.cpp
In file included from ./my-glaze-test-float.cpp:9:
In file included from ./include/glaze/glaze.hpp:35:
In file included from ./include/glaze/binary.hpp:7:
In file included from ./include/glaze/binary/ptr.hpp:6:
In file included from ./include/glaze/binary/read.hpp:7:
In file included from ./include/glaze/binary/skip.hpp:8:
In file included from ./include/glaze/core/read.hpp:9:
In file included from ./include/glaze/core/common.hpp:26:
In file included from ./include/glaze/util/validate.hpp:10:
./include/glaze/core/write_chars.hpp:88:33: error: call to 'to_chars' is ambiguous
               const auto end = glz::to_chars(start, value);
                                ^~~~~~~~~~~~~
./include/glaze/json/write.hpp:195:29: note: in instantiation of function template specialization 'glz::detail::write_chars::op<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, true}, std::vector<unsigned char> &, const float &, glz::context &, unsigned long &>' requested here
               write_chars::op<Opts>(value, ctx, b, ix);
                            ^
./include/glaze/json/write.hpp:30:55: note: in instantiation of function template specialization 'glz::detail::to_json<float>::op<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, true}, std::vector<unsigned char> &, const float &, glz::context &, unsigned long &>' requested here
            to_json<std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
                                                      ^
./include/glaze/json/write.hpp:1324:35: note: in instantiation of function template specialization 'glz::detail::write<10>::op<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, true}, const float &, glz::context &, std::vector<unsigned char> &, unsigned long &>' requested here
                     write<json>::op<opt_true<Opts, &opts::write_unchecked>>(get_member(value, member), ctx, b, ix);
                                  ^
./include/glaze/json/write.hpp:1149:16: note: in instantiation of function template specialization 'glz::detail::to_json<MyStruct>::op_base<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, std::vector<unsigned char> &, const MyStruct &, glz::context &, unsigned long &>' requested here
               op_base<write_unknown_on<Options>()>(std::forward<V>(value), ctx, b, ix);
               ^
./include/glaze/json/write.hpp:30:55: note: in instantiation of function template specialization 'glz::detail::to_json<MyStruct>::op<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, const MyStruct &, glz::context &, std::vector<unsigned char> &, unsigned long &>' requested here
            to_json<std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
                                                      ^
./include/glaze/core/write.hpp:24:44: note: in instantiation of function template specialization 'glz::detail::write<10>::op<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, const MyStruct &, glz::context &, std::vector<unsigned char> &, unsigned long &>' requested here
      detail::write<Opts.format>::template op<Opts>(std::forward<T>(value), ctx, buffer, ix);
                                           ^
./include/glaze/core/write.hpp:54:7: note: in instantiation of function template specialization 'glz::write<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, const MyStruct &, std::vector<unsigned char>, glz::context &>' requested here
      write<Opts>(std::forward<T>(value), buffer, ctx);
      ^
./include/glaze/json/write.hpp:1554:14: note: in instantiation of function template specialization 'glz::write<{10, false, true, true, true, false, false, 32, 3, true, false, true, false, false, false, 0, 0, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, const MyStruct &, std::vector<unsigned char>>' requested here
      return write<opts{}>(std::forward<T>(value), std::forward<Buffer>(buffer));
             ^
./include/glaze/util/itoa.hpp:51:17: note: candidate function [with buf:auto = unsigned char]
   inline auto* to_chars(auto* buf, uint32_t val) noexcept
                ^
./include/glaze/util/itoa.hpp:131:17: note: candidate function [with buf:auto = unsigned char]
   inline auto* to_chars(auto* buf, int32_t x) noexcept
                ^
./include/glaze/util/itoa.hpp:245:17: note: candidate function [with buf:auto = unsigned char]
   inline auto* to_chars(auto* buf, uint64_t val) noexcept
                ^
./include/glaze/util/itoa.hpp:270:17: note: candidate function [with buf:auto = unsigned char]
   inline auto* to_chars(auto* buf, int64_t x) noexcept
                ^
1 error generated.
```

</details>

<details><summary>Actual error log (`g++-12`)</summary>

```
~/glaze-issue-float$ g++-12 -std=c++20 -I ./include ./my-glaze-test-float.cpp
In file included from ./include/glaze/util/validate.hpp:10,
                 from ./include/glaze/core/common.hpp:26,
                 from ./include/glaze/core/read.hpp:9,
                 from ./include/glaze/binary/skip.hpp:8,
                 from ./include/glaze/binary/read.hpp:7,
                 from ./include/glaze/binary/ptr.hpp:6,
                 from ./include/glaze/binary.hpp:7,
                 from ./include/glaze/glaze.hpp:35,
                 from ./my-glaze-test-float.cpp:9:
./include/glaze/core/write_chars.hpp: In instantiation of ‘static void glz::detail::write_chars::op(auto:198&&, auto:199&&, B&&, auto:200&&) [with auto Opts = glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, true}; B = std::vector<unsigned char>&; auto:198 = const float&; auto:199 = glz::context&; auto:200 = long unsigned int&]’:
./include/glaze/json/write.hpp:195:37:   required from ‘static void glz::detail::to_json<T>::op(auto:666&&, auto:667&&, B&&, auto:668&&) [with auto Opts = glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, true}; B = std::vector<unsigned char>&; auto:666 = const float&; auto:667 = glz::context&; auto:668 = long unsigned int&; T = float]’
./include/glaze/json/write.hpp:30:63:   required from ‘static void glz::detail::write<10>::op(T&&, Ctx&&, B&&, IX&&) [with auto Opts = glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, true}; T = const float&; Ctx = glz::context&; B = std::vector<unsigned char>&; IX = long unsigned int&]’
./include/glaze/json/write.hpp:1324:77:   required from ‘glz::detail::to_json<MyStruct>::op_base<glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, std::vector<unsigned char>&, const MyStruct&, glz::context&, long unsigned int&>(const MyStruct&, glz::context&, std::vector<unsigned char>&, long unsigned int&)::<lambda(auto:746)> [with auto:746 = std::integral_constant<long unsigned int, 0>]’
./include/glaze/util/for_each.hpp:16:12:   required from ‘glz::for_each<1, detail::to_json<MyStruct>::op_base<glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, std::vector<unsigned char>&, const MyStruct&, glz::context&, long unsigned int&>(const MyStruct&, glz::context&, std::vector<unsigned char>&, long unsigned int&)::<lambda(auto:746)> >(detail::to_json<MyStruct>::op_base<glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}, std::vector<unsigned char>&, const MyStruct&, glz::context&, long unsigned int&>(const MyStruct&, glz::context&, std::vector<unsigned char>&, long unsigned int&)::<lambda(auto:746)>&&)::<lambda(std::index_sequence<I ...>)> [with long unsigned int ...I = {0}; std::index_sequence<I ...> = std::integer_sequence<long unsigned int, 0>]’
./include/glaze/util/for_each.hpp:17:8:   [ skipping 2 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
./include/glaze/json/write.hpp:1149:52:   required from ‘static void glz::detail::to_json<T>::op(V&&, auto:738&&, auto:739&&, auto:740&&) [with auto Options = glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}; V = const MyStruct&; auto:738 = glz::context&; auto:739 = std::vector<unsigned char>&; auto:740 = long unsigned int&; T = MyStruct]’
./include/glaze/json/write.hpp:30:63:   required from ‘static void glz::detail::write<10>::op(T&&, Ctx&&, B&&, IX&&) [with auto Opts = glz::opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}; T = const MyStruct&; Ctx = glz::context&; B = std::vector<unsigned char>&; IX = long unsigned int&]’
./include/glaze/core/write.hpp:24:52:   required from ‘void glz::write(T&&, Buffer&, auto:360&&) [with opts Opts = opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}; T = const MyStruct&; Buffer = std::vector<unsigned char>; auto:360 = context&]’
./include/glaze/core/write.hpp:54:18:   required from ‘void glz::write(T&&, Buffer&) [with opts Opts = opts{10, false, true, true, true, false, false, ' ', 3, true, false, true, false, false, false, 0, glz::float_precision::full, false, false, false, false, false, false, true, true, false, false, false, false, true, false, false, false}; T = const MyStruct&; Buffer = std::vector<unsigned char>]’
./include/glaze/json/write.hpp:1554:27:   required from ‘auto glz::write_json(T&&, Buffer&&) [with T = const MyStruct&; Buffer = std::vector<unsigned char>&]’
./my-glaze-test-float.cpp:22:20:   required from here
./include/glaze/core/write_chars.hpp:88:46: error: call of overloaded ‘to_chars(unsigned char* const&, const float&)’ is ambiguous
   88 |                const auto end = glz::to_chars(start, value);
      |                                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~
In file included from ./include/glaze/util/dtoa.hpp:8,
                 from ./include/glaze/core/write_chars.hpp:11:
./include/glaze/util/itoa.hpp:51:17: note: candidate: ‘auto* glz::to_chars(auto:135*, uint32_t) [with auto:135 = unsigned char; uint32_t = unsigned int]’
   51 |    inline auto* to_chars(auto* buf, uint32_t val) noexcept
      |                 ^~~~~~~~
./include/glaze/util/itoa.hpp:131:17: note: candidate: ‘auto* glz::to_chars(auto:136*, int32_t) [with auto:136 = unsigned char; int32_t = int]’
  131 |    inline auto* to_chars(auto* buf, int32_t x) noexcept
      |                 ^~~~~~~~
./include/glaze/util/itoa.hpp:245:17: note: candidate: ‘auto* glz::to_chars(auto:141*, uint64_t) [with auto:141 = unsigned char; uint64_t = long unsigned int]’
  245 |    inline auto* to_chars(auto* buf, uint64_t val) noexcept
      |                 ^~~~~~~~
./include/glaze/util/itoa.hpp:270:17: note: candidate: ‘auto* glz::to_chars(auto:142*, int64_t) [with auto:142 = unsigned char; int64_t = long int]’
  270 |    inline auto* to_chars(auto* buf, int64_t x) noexcept
      |                 ^~~~~~~~
```

</details>

- It works fine with other basic primitives such as `byte`, ..., `int32_t`, ..., `uint64_t` etc.  But `float` and `double` are not working properly.
- Since README.md states `clang++-15` and `g++-12` as supported compilers, I also suspect my environment issue.

**Comment by stephenberry** [comment]
@t-mat 
This was actually simpler than I thought it might be, so I was able to fix it right away in . Thanks again for reporting this issue, it's been added as a unit test.

One note of warning is that because you append a null character to your serialized buffer, when you copy to your output string and use `ser.size()`, this will include your null character in your output string, you'll get two null characters at the end of your string and your string's size will include the first null character. It's usually better to simply use `std::string` as your buffers to avoid these kinds of bugs and not need to append your own null characters.

**Comment by stephenberry** [comment]
This has now been fully fixed as on #1011. The floating point conversion code hadn't supported `std::vector<std::byte>` so the float was being casted to an int. Glaze now blocks these implicit conversions and correctly supports `std::vector<std::byte>`
